src/iter.rs

```diff
@@ -1147,3 +1200,132 @@ where
         self.iter.into_iter().flat_map(self.f)
     }
 }
+
+/// Convenience trait extending [`Iterator`]
+pub trait IteratorExt {
+    /// The type of the elements being iterated over.
+    type Item;
+
+    /// Which kind of [`NonEmptyIterator`] are we turning this into?
+    type IntoIter: NonEmptyIterator<Item = Self::Item>;
+
+    /// Tries to convert [`self`] into [`NonEmptyIterator`].
+    fn try_into_nonempty_iter(self) -> Option<Self::IntoIter>;
```

Comment:
This is a great idea, especially given the "Caveat" section in the crate docs:

> Since NEVec, NEMap, and NESet must have a least one element, it is not possible to implement the [FromIterator](https://doc.rust-lang.org/1.74.0/core/iter/traits/collect/trait.FromIterator.html) trait for them. We can’t know, in general, if any given standard-library [Iterator](https://doc.rust-lang.org/1.74.0/core/iter/traits/iterator/trait.Iterator.html) actually contains something.

This allows a user with any given `Iterator` instance to at least try to convert to the non-empty variant.

What do you think about the prefix `try_`? `try_` in the stdlib often (always?) implies a return type of `Result`, but here you're returning `Option`. This crate has no other `Error` type, and introducing one just for the impl below seems a bit silly, so I'm not sure what to do here. Thoughts?

Reply:
Yeah, I was worried about that too. I kind of like naming functions that return on `Option` `maybe_`, a bit like in haskell. I don't really see this in rust so I went with the usual `try_`. Do you like the idea of calling it `maybe_into_nonempty_iter`?

Reply:
Looking for evidence in the stdlib, we can see:
- [`OsStr::to_str`](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#method.to_str), which yields Option.
- `as_*` function always seem to succeed (and just borrow)
- `into_*` functions always seem to succeed (and take ownership)

Let's go with `to_nonempty_iter`.